added cli option `--ignore-errors`

benf710 opened the pull request:
This addresses #54. I wasn't sure on the best way to write the test cases for this. If you can give me some guidance on that I would gladly add them as well.

evfro commented:
Hi! As an example, `psutil` package can be used, and any packages depending on it. I described the case with `wandb` in https://github.com/wimglenn/johnnydep/issues/74.

Originally tested on Windows, but it's easy to reproduce on a Linux system if it doesn't have gcc.

benf710 commented:
@wimglenn are the test cases broken? I tried running them on master and I got the same errors. My suspicion is that something broke with the `jdtest` pypi package.

wimglenn commented:
I made a cosmetic change in https://github.com/wimglenn/johnnydep/pull/75 and the tests passed fine. So if they're broken, it's broken by these changes.

benf710 commented:
If I run test_cli.py `test_printed_table_on_stdout` I get the following error. It seems to indicate that I can't download jdtest. If I search pypi for jdtest, I get no results. Am I doing something wrong here?
```
C:\Users\ben-f\.virtualenvs\johnnydep\Scripts\python.exe "C:\Program Files\JetBrains\PyCharm 2021.2.3\plugins\python\helpers\pycharm\_jb_pytest_runner.py" --target tests/test_cli.py::test_printed_table_on_stdout
Testing started at 8:52 PM ...
Launching pytest with arguments tests/test_cli.py::test_printed_table_on_stdout --no-header --no-summary -q in E:\Users\Ben\Development\johnnydep

============================= test session starts =============================
collecting ... collected 1 item

tests/test_cli.py::test_printed_table_on_stdout FAILED                   [100%]
tests\test_cli.py:8 (test_printed_table_on_stdout)
mocker = <pytest_mock.plugin.MockerFixture object at 0x000002A10B468BB0>
capsys = <_pytest.capture.CaptureFixture object at 0x000002A10B3FAA30>
make_dist = <function make_dist.<locals>.f at 0x000002A10B3F5F70>

    def test_printed_table_on_stdout(mocker, capsys, make_dist):
        make_dist()
        mocker.patch("sys.argv", "johnnydep jdtest==0.1.2".split())
>       main()

tests\test_cli.py:12: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
johnnydep\cli.py:77: in main
    dist = JohnnyDist(
johnnydep\lib.py:64: in __init__
    self.import_names, self.metadata = _get_info(
C:\Users\ben-f\.virtualenvs\johnnydep\lib\site-packages\cachetools\func.py:61: in wrapper
    v = func(*args, **kwargs)
johnnydep\lib.py:392: in _get_info
    data = pipper.get(
C:\Users\ben-f\.virtualenvs\johnnydep\lib\site-packages\cachetools\__init__.py:520: in wrapper
    v = func(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

dist_name = 'jdtest==0.1.2', index_url = None, env = None
extra_index_url = None
tmpdir = 'C:\\Users\\ben-f\\AppData\\Local\\Temp\\tmp9kvf5wrd'

    @cached(_get_cache, key=_cache_key)
    def get(dist_name, index_url=None, env=None, extra_index_url=None, tmpdir=None):
        args = _get_wheel_args(index_url, env, extra_index_url) + [dist_name]
        scratch_dir = tempfile.mkdtemp(dir=tmpdir)
        log.debug("wheeling and dealing", scratch_dir=os.path.abspath(scratch_dir), args=" ".join(args))
        try:
            out = subprocess.check_output(args, stderr=subprocess.STDOUT, cwd=scratch_dir)
        except subprocess.CalledProcessError as err:
            output = getattr(err, "output", b"").decode("utf-8")
            log.warning(output)
            raise
        log.debug("wheel command completed ok", dist_name=dist_name)
        out = out.decode("utf-8")
        links = []
        lines = out.splitlines()
        for i, line in enumerate(lines):
            line = line.strip()
            if line.startswith("Downloading from URL "):
                parts = line.split()
                link = parts[3]
                links.append(link)
            elif line.startswith("Downloading "):
                parts = line.split()
                last = parts[-1]
                if len(parts) == 3 and last.startswith("(") and last.endswith(")"):
                    link = parts[-2]
                elif len(parts) == 4 and parts[-2].startswith("(") and last.endswith(")"):
                    link = parts[-3]
                    if not urlparse(link).scheme:
                        # newest pip versions have changed to not log the full url
                        # in the download event. it is becoming more and more annoying
                        # to preserve compatibility across a wide range of pip versions
                        next_line = lines[i + 1].strip()
                        if next_line.startswith("Added ") and " to build tracker" in next_line:
                            link = next_line.split(" to build tracker")[0].split()[-1]
                else:
                    link = last
                links.append(link)
            elif line.startswith("Source in ") and "which satisfies requirement" in line:
                link = line.split()[-1]
                links.append(link)
        links = list(OrderedDict.fromkeys(links))  # order-preserving dedupe
        if not links:
            log.warning("could not find download link", out=out)
>           raise Exception("failed to collect dist")
E           Exception: failed to collect dist

johnnydep\pipper.py:181: Exception

```

wimglenn commented:
`jdtest` doesn't exist on PyPI.  It is created during test setup and only exists locally.

benf710 commented:
Any help or guidance you can give me on this would be much appreciated. Maybe there is an issue with a dependency version that I'm using? If I clone the master branch and run pytest I get many failures:

```PS E:\Users\Ben\Development\test-johnny-dep> git clone https://github.com/wimglenn/johnnydep.git
Cloning into 'johnnydep'...
remote: Enumerating objects: 437, done.
remote: Counting objects: 100% (80/80), done.
remote: Compressing objects: 100% (61/61), done.
Receiving objects:  97% (424/437)sed 42 (delta 17), pack-reused 357R
Receiving objects: 100% (437/437), 345.38 KiB | 1.35 MiB/s, done.
Resolving deltas: 100% (240/240), done.
PS E:\Users\Ben\Development\test-johnny-dep> cd johnny-dep
PS E:\Users\Ben\Development\test-johnny-dep\johnnydep> python -m venv .venv
PS E:\Users\Ben\Development\test-johnny-dep\johnnydep> .\.venv\Scripts\activate
(.venv) PS E:\Users\Ben\Development\test-johnny-dep\johnnydep> pip -V
pip 19.2.3 from e:\users\ben\development\test-johnny-dep\johnnydep\.venv\lib\site-packages\pip (python 3.8)
(.venv) PS E:\Users\Ben\Development\test-johnny-dep\johnnydep> pip install -r .\requirements-dev.txt
Obtaining file:///E:/Users/Ben/Development/test-johnny-dep/johnnydep (from -r .\requirements-dev.txt (line 1))
...
Successfully installed anytree-2.8.0 atomicwrites-1.4.0 attrs-21.2.0 cachetools-4.2.4 colorama-0.4.4 coverage-6.2 distlib-0.3.3 iniconfig-1.1.1 johnnydep oyaml-1.0 packaging-21.3 pkginfo-1.8.1 pluggy-1.0.0 py-1.11.0 pyparsing-3.0.6 pytest-6.2.5 pytest-cov-3.0.0 pytest-mock-3.6.1 pytest-socket-0.4.1 pyyaml-6.0 six-1.16.0 structlog-21.4.0 tabulate-0.8.9 toml-0.10.2 wheel-0.37.0 wimpy-0.3
WARNING: You are using pip version 19.2.3, however version 21.3.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.
(.venv) PS E:\Users\Ben\Development\test-johnny-dep\johnnydep> pip list installed
Package       Version Location
------------- ------- --------------------------------------------------
anytree       2.8.0
atomicwrites  1.4.0
attrs         21.2.0
cachetools    4.2.4
colorama      0.4.4
coverage      6.2
distlib       0.3.3
iniconfig     1.1.1
johnnydep     1.10    e:\users\ben\development\test-johnny-dep\johnnydep
oyaml         1.0
packaging     21.3
pip           19.2.3
pkginfo       1.8.1
pluggy        1.0.0
py            1.11.0
pyparsing     3.0.6
pytest        6.2.5
pytest-cov    3.0.0
pytest-mock   3.6.1
pytest-socket 0.4.1
PyYAML        6.0
setuptools    41.2.0
six           1.16.0
structlog     21.4.0
tabulate      0.8.9
toml          0.10.2
wheel         0.37.0
wimpy         0.3
WARNING: You are using pip version 19.2.3, however version 21.3.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.
(.venv) PS E:\Users\Ben\Development\test-johnny-dep\johnnydep> python -m pytest
================================================================================================================================= test session starts =================================================================================================================================
platform win32 -- Python 3.8.0, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: E:\Users\Ben\Development\test-johnny-dep\johnnydep, configfile: pytest.ini, testpaths: tests
plugins: cov-3.0.0, mock-3.6.1, socket-0.4.1
collected 79 items

tests\test_cli.py FFFFFFF                                                                                                                                                                                                                                                        [  8%]
tests\test_gen.py F.FFFFF                                                                                                                                                                                                                                                        [ 17%]
tests\test_lib.py FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF.FFFF                                                                                                                                                                                                         [ 86%]
tests\test_pipper.py F.......                                                                                                                                                                                                                                                    [ 96%]
tests\test_util.py ...      ```

wimglenn commented:
Oh, you are using Windows. I'll take a look at what goes wrong on windows, because all my CI runs on Linux.

benf710 commented:
I didn't think that would be the issue since my PR failed in the GitHub actions CI but I re-ran the tests in an Ubuntu VM and they all passed...

wimglenn commented:
I don't really develop on Windows so if johnnydep worked on windows it was pure luck  :) 
But I've added macOS and Windows to the matrix in the tests  #76  and fixed some things which were accidentally platform-dependent

benf710 commented:
@wimglenn let me know if anything else is needed for this. This test might not be great but it seems that the error message changes depending on the version of Python: https://github.com/wimglenn/johnnydep/blob/af8bd8ab60f3775fabd626dd000a617190fe5eef/tests/test_cli.py#L177

wimglenn commented on tests/test_lib.py:
```diff
@@ -510,5 +526,6 @@ def test_license_parsing_unknown(make_dist):
 def test_metadata_cant_be_extracted(make_dist, mocker):
     make_dist()
     mocker.patch("pkginfo.get_metadata", return_value=None)
-    with pytest.raises(Exception("failed to get metadata")):
+    with pytest.raises(Exception) as ex:
         JohnnyDist("jdtest")
+    assert str(ex.value) == "failed to get metadata"
```
Why modifying this test?  You may have forgot to install a [new development dependency](https://github.com/wimglenn/johnnydep/blob/71b51d389ef28cc22b84a5c3f46e4f635b4d99c0/requirements-dev.txt#L5) that was added in #76

wimglenn commented on johnnydep/lib.py:
```diff
@@ -39,14 +40,38 @@ def __missing__(self, key):
+            self.error = lines[-3]
+        except (ValueError, AttributeError):
+            self.error = "Exception: Unable to parse exception message"
+        super(JohnnyDistFailed, self).__init__(parent)
```
This class is missing many of the attributes that a `JohnnyDist` instance needs to have (stuff that normally would be set in `JohnnyDist` init).  For example when triggering a failure with `johnnydep wandb --ignore-errors -o yaml` I get:

```
Traceback (most recent call last):
  File "C:\Program Files\Python310\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Program Files\Python310\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\wglenn\git\johnnydep\johnnydep\__main__.py", line 4, in <module>
    main()
  File "C:\Users\wglenn\git\johnnydep\johnnydep\cli.py", line 85, in main
    print(dist.serialise(fields=args.fields, format=args.output_format, recurse=args.recurse))
  File "C:\Users\wglenn\git\johnnydep\johnnydep\lib.py", line 282, in serialise
    next(deps)  # skip over root
  File "C:\Users\wglenn\git\johnnydep\johnnydep\lib.py", line 339, in flatten_deps
    extra_map[dep.name] |= set(dep.extras_requested)
AttributeError: 'JohnnyDistFailed' object has no attribute 'extras_requested'
```

I think you should try to implement this feature by handling errors within `JohnnyDist` directly, without splitting into two separate types.

benf710 replied:
Thanks for reviewing. I have no issue re-writing this but I'm curious why you think it would be better to handle the errors within JohnnyDist instead of with inheritance? My initial thought was that one might want to distinguish between a successful and a failed JonnyDist type.

wimglenn replied:
They can be distinguished with an attribute.  They're not so different enough that distinct types are necessary.

benf710 commented:
I've re-written this, however I'm running into an issue with TTLCache in [pipper.get()](https://github.com/benf710/johnnydep/blob/6d8b45decaf1719f68c614bf6e211ead196c844b/johnnydep/pipper.py#L138). 

**When I remove the caching** for this method everything works fine, but if I leave it alone I get a type error (seems to be related to the extra argument in the function call?).  Wondering if you've seen anything like this before or have any ideas for what I should do?

```
test_cli.py::test_printed_table_on_stdout FAILED                         [100%]
tests\test_cli.py:8 (test_printed_table_on_stdout)
mocker = <pytest_mock.plugin.MockerFixture object at 0x000001CC4BAD9490>
capsys = <_pytest.capture.CaptureFixture object at 0x000001CC4BACFC70>
make_dist = <function make_dist.<locals>.f at 0x000001CC4BAD3940>

    def test_printed_table_on_stdout(mocker, capsys, make_dist):
        make_dist()
        mocker.patch("sys.argv", "johnnydep jdtest==0.1.2".split())
>       main()

test_cli.py:12: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\johnnydep\cli.py:79: in main
    dist = JohnnyDist(
..\johnnydep\lib.py:66: in __init__
    self.import_names, self.metadata = _get_info(
C:\Users\ben-f\.virtualenvs\johnnydep\lib\site-packages\cachetools\func.py:61: in wrapper
    v = func(*args, **kwargs)
..\johnnydep\lib.py:406: in _get_info
    data = pipper.get(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = ('jdtest==0.1.2',)
kwargs = {'env': None, 'extra_index_url': None, 'ignore_errors': False, 'index_url': None, ...}

    def wrapper(*args, **kwargs):
>       k = key(*args, **kwargs)
E       TypeError: _cache_key() got an unexpected keyword argument 'ignore_errors'

C:\Users\ben-f\.virtualenvs\johnnydep\lib\site-packages\cachetools\__init__.py:515: TypeError
```

wimglenn commented:
@benf710 hard to say without seeing what's changed - push up the changes?

benf710 commented:
Code has been pushed to my fork: 
https://github.com/benf710/johnnydep/tree/feat/ignore-errors-no-inheritance 

CI Results: 
`TypeError: _cache_key() got an unexpected keyword argument 'ignore_errors'`
https://github.com/benf710/johnnydep/runs/4687032550?check_suite_focus=true

benf710 commented:
Found it, looks like I missed the function [_cache_key:128](https://github.com/benf710/johnnydep/blob/6d8b45decaf1719f68c614bf6e211ead196c844b/johnnydep/pipper.py#L128).

benf710 commented:
Created the re-write PR in #77
